compiler/rustc_lint/src/dangling.rs

```diff
@@ -199,24 +200,25 @@ fn is_temporary_rvalue(expr: &Expr<'_>) -> bool {
     }
 }
 
-// Array, Vec, String, CString, MaybeUninit, Cell, Box<[_]>, Box<str>, Box<CStr>,
-// or any of the above in arbitrary many nested Box'es.
-fn is_interesting(tcx: TyCtxt<'_>, ty: Ty<'_>) -> bool {
+// Array, Vec, String, CString, MaybeUninit, Cell, Box<[_]>, Box<str>, Box<CStr>, UnsafeCell,
+// SyncUnsafeCell, or any of the above in arbitrary many nested Box'es.
+fn owns_allocation(tcx: TyCtxt<'_>, ty: Ty<'_>) -> bool {
```

Comment:
With the addition of the attribute we should able to simplify this `owns_allocation` method to a simple check of the type and it's generic arguments.

Something like this would be a good starting point (*the check should probably be further refined*):
```rust
    match ty.kind() {
        ty::Array(_, _) => true,
        ty::Adt(_def, args) => args.iter().filter_map(|g| g.as_type()).all(|ty| !ty.is_any_ptr()),
        _ => false,
    }
```

Reply:
I think a `Box<Box<&[T]>>` would pass the check that you suggest, even though it should not

Reply:
Yeah, we still need to take the derefs into account.

<details>

Probably something more like this:
```rust
    let adjs = cx.typeck_results().expr_adjustments(receiver);
    let origin_ty = cx.typeck_results().expr_ty(receiver);

    adjs.iter()
        .filter_map(|adj| match adj.kind {
            Adjust::Deref(_) => Some(adj.target),
            Adjust::NeverToAny
            | Adjust::Borrow(_)
            | Adjust::Pointer(_)
            | Adjust::ReborrowPin(_) => None,
        })
        .all(|ty| !ty.is_any_ptr())
        && match origin_ty.kind() {
            ty::Adt(_def, args) => {
                args.iter().filter_map(|g| g.as_type()).all(|ty| !ty.is_any_ptr())
            }
            _ => !origin_ty.is_any_ptr(),
        }
```

but, let's discussed that on the follow-up issue/pr.

</details>